Zotero/Controllers/OpenItemsController.swift

```diff
@@ -0,0 +1,481 @@
+                    token = objects.observe { [weak self] changes in
+                        switch changes {
+                        case .initial:
+                            break
+
+                        case .update(_, let deletions, _, _):
+                            if !deletions.isEmpty, let self {
+                                // Observed items have been deleted, call setItems to validate and register new observer.
+                                let existingItems = getItems(for: sessionIdentifier)
+                                setItems(existingItems, for: sessionIdentifier, validate: true)
+                            }
+
+                        case .error(let error):
+                            DDLogError("OpenItemsController: register observer error - \(error)")
+                        }
+                    }
+                }
+            } catch let error {
+                DDLogError("OpenItemsController: can't register items observer - \(error)")
+            }
+            return token
+        }
+    }
+
+    private func setItemsSortedByUserIndex(_ items: [Item], for sessionIdentifier: String, validate: Bool) {
+        var newItems = items
+        for i in 0..<newItems.count {
+            newItems[i].userIndex = i
+        }
+        setItems(newItems, for: sessionIdentifier, validate: validate)
+    }
+
+    func sessionIdentifier(for kind: Item.Kind) -> String? {
+        sessionIdentifierByItemKind[kind]
+    }
+
+    func open(_ kind: Item.Kind, for sessionIdentifier: String) {
+        DDLogInfo("OpenItemsController: opened item \(kind) for \(sessionIdentifier)")
+        var existingItems = getItems(for: sessionIdentifier)
+        if let index = existingItems.firstIndex(where: { $0.kind == kind }) {
+            existingItems[index].lastOpened = .now
+            // No need to call setItems, to register a new items observer, as only items metadata were updated.
+            itemsBySessionIdentifier[sessionIdentifier] = existingItems
+            DDLogInfo("OpenItemsController: already opened item \(kind) became most recent for \(sessionIdentifier)")
+            observable(for: sessionIdentifier).on(.next(existingItems))
+        } else {
+            DDLogInfo("OpenItemsController: newly opened item \(kind) set as most recent for \(sessionIdentifier)")
+            let item = Item(kind: kind, userIndex: existingItems.count)
+            let newItems = existingItems + [item]
+            // setItems will produce next observable event
+            setItems(newItems, for: sessionIdentifier, validate: false)
+        }
+    }
+
+    func close(_ kind: Item.Kind, for sessionIdentifier: String) {
+        DDLogInfo("OpenItemsController: closed open item \(kind) for \(sessionIdentifier)")
+        var existingItems = itemsSortedByUserOrder(for: sessionIdentifier)
+        if let index = existingItems.firstIndex(where: { $0.kind == kind }) {
```

Comment:
This would look nicer as a guard, but not mandatory. Same with `move` function below.

Reply:
Check https://github.com/zotero/zotero-ios/pull/829/commits/ae2376422f3f11d32545767dac3d7b35913fac7a